docs/reference/config/ui-options.md
---
description: These are the options for controlling the UI of the program
---

# `mode`

| Type                                            | Default      |
| ----------------------------------------------- | ------------ |
| `DISABLED`, `ACTIVITY`,`SIMPLE` or `FULLSCREEN` | `FULLSCREEN` |

- `DISABLED` : no output at all, only log messages (if enabled)
- `ACTIVITY` : only shows a spinner with the text `running cyberdrop-dl vX.Y.Z...`
- `SIMPLE`: shows spinner + simplified progress bar
- `FULLSCREEN`: shows the normal UI/progress view

```yaml
ui:
  mode: fullscreen
```

{% hint style="info" %}
Values are case insensitive, ex: both `disabled` and `DISABLED` are valid
{% endhint %}

# `portrait`

| Type   | Default |
| ------ | ------- |
| `Bool` | `False` |

Force CDL to run with a vertical layout

```yaml
ui:
  portrait: false
```

# `refresh_rate`

| Type            | Default |
| --------------- | ------- |
| `PositiveFloat` | `10.0`  |

Refresh the UI <n> times per second. Changing this value may help with screen flickering

```yaml
ui:
  refresh_rate: 10.0
```

# `show_stats`

| Type   | Default |
| ------ | ------- |
| `Bool` | `true`  |

Show stats report at the end of a run

```yaml
ui:
  show_stats: true
```


CHANGELOG.md
<!-- Template
## UNRELEASED

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

-->

# CHANGELOG

All notable changes to this project will be documented here. For more details, visit the wiki: <https://script-ware.gitbook.io/cyberdrop-dl>

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## UNRELEASED

### Added

- New `--flaresolverr-wait` option
- New `--flaresolverr-concurrency` option
- Support file paths as positional arguments
- Support multiple files as input
- Support mixing URLs and input files on the same run (all files must exists and all URLs/files must be positional arguments)

### Changed

- Allow multiple concurrent requests with Flaresolverr
- `--skip-hosts` and `--only-hosts` now perform exact domain matching if an absolute URL is provided.
  ex: `--skip-hosts https://x.com` will skip `x.com` URLs but not `vix.com` URLs. Using `--skip-hosts x.com` will skip both
- Hardcoded max rate limit to 3 requests/second (Pawchive)
- Hardcoded concurrent downloads limit to 5 (Pawchive)

### Fixed

- Corrupted/mixed up video segments when downloading multiple concurrent HLS streams and `--subfolders.create` is `False`
- `403 Forbidden` on all downloads (AdobeLightroom)
- `400 Bad request` for albums (PornHub)
- URL matching (Anysex)

## [10.7.0] - 2026-08-29

### Added

- New `--downloads.back-pressure` option
- New `--restrict-path` option
- New `--input-folder` option that scans all `.txt` within a folder for URLs (non recursive). URLs within the same file will create a group. The name of the group is the filename;
  Each group will be downloaded to their own subfolder within the downloads folder. See: <https://script-ware.gitbook.io/cyberdrop-dl/reference/config/sorting#group-urls>
- New `--input` option that accepts either a file or a folder
- Clonr.co support

### Changed

- `-i` is now an alias for `--input` instead of `--input-file`
- Increase limit of concurrent scrapes from the same site, from 20 to 40
- Use multiple concurrent connections to the database for faster reads (up to 10)
- Database will now operate in WAL mode, which may create additional temp files next to the database file (`.db-shm` and `.db-wal` files)
- Download previews of video assets (Patreon)

### Fixed

- Clips downloads (Twitch)
- Single post downloads (Patreon)
- Download of image assets with no name (Patreon)

## [10.6.1] - 2026-08-25

### Changed

- Hashing files no longer blocks the download queue. When a download reaches 100%, a slot is immediately available while hashes are computed in the background (if enabled)
- Restore hardcoded limit of 1 concurrent download per server (Bunkr)

### Fixed

- All downloads freezing if too many files were currently hashing

## [10.6.0] - 2026-08-24

### Added

- New `--logs.http-traffic` option
- New `--filename-regex-exclude` option
- New `--flaresolverr-use-session` option
- Support direct local connection to JDownloader using their deprecated API
- `wreq` as optional HTTP backend for browser impersonation (via the `wreq` extra)
- Webmshare support
- Support for revisions (Pawchive)

### Changed

- Log HTTP requests made to JDownloader
- Logs of HTTP requests/responses will be emitted at INFO instead of DEBUG level
- If scraping has already started, sending `SIGINT` (`Ctrl + C`) to the process will stop all scraping and downloads but will continue with post-runtime actions
  (sorting, hashing, send notifications, etc..). Press `Ctrl + C` again to force quit
- Reduced the maximum size the log file can have to be sent as attachment in notifications, from 25 MB to 20 MB.
- Retry downloads (from 0%) on transient network errors (Mega.nz)

### Fixed

- All scrapes failing if running with `--dump-responses`
- Handle malformed JSON responses wrapped in HTML (Flaresolverr)
- CLI additive args being ignored
- `PermissionError` running `database transfer`
- Incorrent warning about login cookies (vBulletin)
- Handle new signed download URLs (Fileditch)
- Handle new URL format (Whyp.it)
- `416 - Precondition failed` on all downloads (PornHub)

## [10.5.0] - 2026-08-15

### Added

- Support defining config settings across 2 different files.
  A second config file can be created at the same location as the current `--config-file` with the suffix (`.override`):
  ex: `/.config/cyberdrop-dl/config.yaml`, `/.config/cyberdrop-dl/config.override.yaml`.

### Fixed

- Update website salt (Gofile)

## [10.4.0] - 2026-08-14

### Added

- New `--ignore-hashes` option
- New `--thumbnails` option
- New `--tls.min-version` option
- New `--ca-certs` option
- New `--verify` option
- New `database dir` command
- New `config dir` command
- New `cache dir` command
- New generic `video` crawler to map any unsupported site as a video site
- Support for HLS streams with encrypted segments (AES-128)
- Support setting a custom referer for some sites by setting the `referer` query param to the referer URL
- Vidstack support
- Pluto.tv support
- Nova.storage support
- OnlyHaven support
- Odysee support
- Hohoj support
- Mitaku support
- Octave Music support
- Redtube support
- Livecamrips support
- xpornium support
- Gifhq support
- Fyptt support
- Short URLs support (Filedicth)
- Users support (Goonbox)
- Hashtags support (TWPornstars)

### Changed

- Try to download M3U8 URLs from unsupported sites
- Always use the CA certificates from the system trust store + CA bundle from Mozilla
- Skip all files with pending imports (Pawchive)
- The full URL of successful downloads from unsupported sites (`no_crawler`) will be saved as is to the database. Existing database entries are invalid
- Always include content id in folder name if its name is `root` (GoFile)

cyberdrop_dl/config/settings.py
import datetime
import logging
import random
from enum import auto
from pathlib import Path
from typing import Annotated, Any, ClassVar, Literal, override

from cyclopts import Parameter
from pydantic import BaseModel, Field, NonNegativeInt, PrivateAttr
from pydantic.functional_validators import BeforeValidator
from pydantic.types import ByteSize, NonNegativeFloat, PositiveFloat, PositiveInt

from cyberdrop_dl.constants import LOGS_DATE_FORMAT, LOGS_DATETIME_FORMAT, CIStrEnum, HashMode, ImpersonateTarget
from cyberdrop_dl.models import ConfigGroup, ConfigModel
from cyberdrop_dl.models.types import (
    ByteSizeSerilized,
    CSVPath,
    ExistingPath,
    FalsyAsNone,
    FormatStr,
    HttpURL,
    LogLevel,
    LogPath,
    NonEmptyStr,
    Timedelta,
)
from cyberdrop_dl.models.validators import strings

ALIASES: dict[str, str] = {}


def _alias(name: str, *extras: str, flag: str = "enabled") -> Parameter:
    aliases = (f"--{name}", f"--{name}.{flag}")
    negative_aliases = (f"--{name}.no-{flag}", f"--no-{name}")
    ALIASES.update([aliases, negative_aliases])
    return Parameter(alias=aliases + extras, negative_alias=negative_aliases)


class _SubFoldersInclude(ConfigModel):
    album_id: bool = False
    thread_id: bool = False
    domain: bool = True


@Parameter(name="*")
class _SeparatePosts(ConfigModel):
    format: Annotated[
        FormatStr,
        strings.format_validator({"default", "title", "id", "number", "date"}),
        Parameter(name="separate-posts.format"),
    ] = "{default}"

    enabled: Annotated[bool, _alias("separate-posts")] = False
    "Create new subfolders for every post on a site"


class SubFolders(ConfigGroup, name=None):
    create: Annotated[bool, _alias("subfolders", flag="create")] = True
    "Enable/disable the createtion of nested sub-folders"

    include: _SubFoldersInclude = Field(default_factory=_SubFoldersInclude)
    separate_posts: _SeparatePosts = Field(default_factory=_SeparatePosts)


class LogFiles(ConfigModel):
    main: Annotated[LogPath, Parameter(alias="--log-file")] = Path("downloader.log")
    "Path of main log file"

    download_errors: CSVPath = Path("download_errors.csv")
    "Save download errors to this file (MUST BE .csv)"

    scrape_errors: CSVPath = Path("scrape_errors.csv")
    "Save scrape errors to this file (MUST BE .csv)"

    unsupported: CSVPath = Path("unsupported.csv")
    "Save unsupported URLs to this file (MUST BE .csv)"

    last_forum_post: CSVPath = Path("last_forum_post.csv")
    "Save the URL of the last scraped post from each thread to this file (MUST BE .csv)"

    dedupe: CSVPath = Path("dedupe.csv")
    "Save every duplicate deleted by the deduper, and the file it matched, to this file (MUST BE .csv)"

    @property
    def jsonl_file(self) -> Path:
        return self.main.with_suffix(".results.jsonl")


class Logs(ConfigGroup, name=None):  # noqa: PLW1641
    level: LogLevel = "DEBUG"
    "Only log messages of this level or higher to the main log file"

    console_level: FalsyAsNone[LogLevel] = None
    "Only log messages of this level or higher to the console. An empty or `None` value will use the same level as `log_level`"

    http_traffic: Annotated[bool, Parameter(alias="--print-traffic")] = True
    "Log HTTP requests and responses (at INFO level)"

    files: LogFiles = Field(default_factory=LogFiles)
    folder: FalsyAsNone[Path] = None
    "Base folder to prepend to log files paths (if they are not absolute)"

    expire_after: FalsyAsNone[Timedelta] = None
    "Delete all log files inside `--logs.folder` if they are older that this"

    rotate: bool = False
    "Append current datetimme to every log file on each run"

    _created_at: datetime.datetime = PrivateAttr(default_factory=datetime.datetime.now)

    @property
    def effective_level(self) -> int:
        return logging.getLevelNamesMapping()[self.level]

    @property
    def effective_console_level(self) -> int:
        if not self.console_level:
            return self.effective_level

        return logging.getLevelNamesMapping()[self.console_level]

    def resolve_filenames(self, appdata_folder: Path) -> None:
        self.folder = folder = self.folder.expanduser().resolve().absolute() if self.folder else appdata_folder
        now_file_iso: str = self._created_at.strftime(LOGS_DATETIME_FORMAT)
        now_folder_iso: str = self._created_at.strftime(LOGS_DATE_FORMAT)

        def resolve(path: Path) -> Path:
            log_file = folder / path
            if self.rotate:
                file_name = f"{log_file.stem}_{now_file_iso}{log_file.suffix}"
                log_file = log_file.parent / now_folder_iso / file_name
            return log_file

        self.files = LogFiles.model_construct(
            None, **{name: resolve(value) for name, value in self.files.model_dump().items()}
        )

    @property
    def effective_log_folder(self) -> Path:
        assert self.folder
        return self.folder

    def delete_old_logs_and_folders(self) -> None:
        if not self.expire_after:
            return

        for file in (self.effective_log_folder).rglob("*"):
            if file.suffix.lower() not in {".log", ".csv"}:
                continue

            if (self._created_at - datetime.datetime.fromtimestamp(file.stat().st_ctime)) > self.expire_after:  # noqa: DTZ006
                file.unlink()

    def __eq__(self, other: object) -> bool:
        # Exclude _created_at from compare (AKA __pydantic_private__)
        if not isinstance(other, BaseModel):
            return NotImplemented

        self_type = self.__pydantic_generic_metadata__["origin"] or self.__class__
        other_type = other.__pydantic_generic_metadata__["origin"] or other.__class__

        if not (self_type == other_type and self.__pydantic_extra__ == other.__pydantic_extra__):
            return False

        return self.__dict__ == other.__dict__


class Jdownloader(ConfigGroup, name=None):
    enabled: Annotated[bool, _alias("jdownloader")] = False
    "Send unsupported URLs to Jdownloader"

    deprecated_api: HttpURL | None = None
    "HTTP URL of a local JDownloader instance to connect to via their deprecated API (insecure, default port=3128)"

    autostart: bool = False
    "Immediately start downloads as soon as they are sent"

    download_folder: FalsyAsNone[Path] = None
    "Output path for Jdownloader. Defaults to `--download-folder`"

    whitelist: set[NonEmptyStr] = Field(default_factory=set)
    "Only send unsupported URLs from these domains to Jdownloader. An empty list means 'send all URLs'"

    @override
    def model_post_init(self, context: Any, /) -> None:
        super().model_post_init(context)
        if self.deprecated_api and self.deprecated_api.scheme != "http":
            raise ValueError("Deprecated API URL must have an 'http' scheme")


class SortFormats(ConfigModel):
    _COMMON_FIELDS: ClassVar[set[str]] = {
        "base_dir",
        "ext",
        "file_date",
        "file_date_iso",
        "file_date_us",
        "filename",
        "parent_dir",
        "sort_dir",
    }

    audio: Annotated[
        FalsyAsNone[FormatStr],
        strings.format_validator(_COMMON_FIELDS | {"bitrate", "duration", "length", "sample_rate"}),
    ] = "{sort_dir}/{base_dir}/Audio/{filename}{ext}"
    "Format to generate sorted audio file"

    image: Annotated[
        FalsyAsNone[FormatStr], strings.format_validator(_COMMON_FIELDS | {"height", "resolution", "width"})
    ] = "{sort_dir}/{base_dir}/Images/{filename}{ext}"
    "Format to generate sorted image file"

    non_media: Annotated[FalsyAsNone[FormatStr], strings.format_validator(_COMMON_FIELDS)] = (
        "{sort_dir}/{base_dir}/Other/{filename}{ext}"
    )
    "Format to generate sorted files of unknown type"

    video: Annotated[
        FalsyAsNone[FormatStr],
        strings.format_validator(
            _COMMON_FIELDS
            | {
                "codec",
                "duration",
                "fps",
                "height",
                "length",
                "resolution",
                "width",
            }
        ),
    ] = "{sort_dir}/{base_dir}/Videos/{filename}{ext}"
    "Format to generate sorted video file"

    incrementer: Annotated[FormatStr, strings.format_validator({"i"})] = " ({i})"
    "Format for separator on name collisions"


class Sort(ConfigGroup, name=None):
    enabled: Annotated[bool, _alias("sort")] = False
    "Enable/Disable file sorting at the end of a run"

    unzip_archives: bool = False
    "Unzip downloaded .zip archives before sorting and delete the original archives"

    input_folder: FalsyAsNone[Path] = None
    "Base folder to scan for files. Default to the same value as `--download-folder`"

    output_folder: Path = Path("downloads/cyberdrop-dl sorted")
    "Output path to place sorted files in"

    formats: SortFormats = Field(default_factory=SortFormats)

    @property
    def needs_ffmpeg(self) -> bool:
        return bool(self.enabled and (self.formats.audio or self.formats.video))


class Dedupe(ConfigModel):
    enabled: Annotated[bool, _alias("dedupe", "--auto-dedupe")] = True
    "Auto delete duplicate downloads by hash"

    use_trash_bin: bool = True
    "Send deleted files to the trash bin"


class Hashing(ConfigGroup, name=None):
    mode: Annotated[HashMode, Parameter(name="--hashing")] = HashMode.IN_PLACE
    algorithms: Annotated[
        tuple[
            Annotated[
                Literal["xxh128", "md5", "sha256"],
                strings.pre_validator(to_lower=True, strip=True),
            ],
            ...,
        ],
        Parameter(alias="--hashes"),
    ] = (
        "xxh128",
        "md5",
        "sha256",
    )
    "List of hashes to compute for each download"

    dedupe: Dedupe = Field(default_factory=Dedupe)
    _extra_hashes: tuple[Literal["md5", "sha256"], ...] = ()

    @override
    def model_post_init(self, context: Any, /) -> None:
        super().model_post_init(context)
        self.re_compute()

    def re_compute(self) -> None:
        hashes = set(self.algorithms)
        if (xxhash := "xxh128") not in hashes:
            self.algorithms = xxhash, *hashes
        hashes.discard(xxhash)
        self._extra_hashes = tuple(sorted(hashes))  # pyright: ignore[reportAttributeAccessIssue]

    @property
    def extra_hashes(self) -> tuple[Literal["md5", "sha256"], ...]:
        return self._extra_hashes


class Downloads(ConfigGroup):
    concurrency: Annotated[PositiveInt, Parameter(name="--downloads")] = 15
    "Max number of files to download simultaneously"

    concurrency_per_domain: Annotated[PositiveInt, Parameter(name="--downloads.per-domain")] = 5
    "Max number of files to download simultaneously per domain"

    attempts: PositiveInt = 2

    delay: NonNegativeFloat = 0.0
    "Number of seconds to wait before starting downloads"

    slow_speed: ByteSizeSerilized = ByteSize(0)
    "Skip downloads if their speed is below this value for more than 10 seconds. Set to 0 to disable"

    speed_limit: ByteSizeSerilized = ByteSize(0)
    "Max speed rate (in bytes per second) to limit downloads (combined)"

    back_pressure: bool = True
    "Throttle scrape requests if there are too many downloads queued for the same site"

    jitter: NonNegativeFloat = 0
    "Wait a random additional number of seconds in between 0 and <jitter> before downloads"

    skip_and_mark_completed: bool = False
    "Skip all downloads and mark them as downloaded on the database"

    concurrent_segments: PositiveInt = 10
    """Allow up to `<N>` HLS segments to be downloaded concurrently"""

    @property
    def total_delay(self) -> NonNegativeFloat:
        return self.delay + random.uniform(0, self.jitter)


@Parameter(name="*")
class TLS(ConfigModel):
    verify: Annotated[bool, Parameter(alias="ssl")] = True
    min_version: Annotated[Literal["1.2", "1.3"], BeforeValidator(str), Parameter(name="tls.min-version")] = "1.2"
    ca_certs: tuple[ExistingPath, ...] = ()


class Network(ConfigGroup):
    dump_responses: bool = False
    "Save text/HTML/JSON responses to disk (Flaresolverr responses are excluded)"

    flaresolverr: FalsyAsNone[HttpURL] = None
    "HTTP URL of an existing Flaresolverr instance"

    flaresolverr_use_session: bool = True
    "Create a custom session before making any request with Flaresolverr"

    flaresolverr_wait: NonNegativeInt = 0
    "Force Flaresolverr to wait (at least) this number of seconds before returning the results, to allow dynamic content to load"

    flaresolverr_concurrency: PositiveInt = 1
    "Number of concurrent requests to make with Flaresolverr"

    playwright: bool = False
    "Use Playwright to bypass Cloudflare/DDOS-Guard challenges automatically"

    playwright_visible: bool = False
    "Show the Playwright browser window instead of running headlessly"

    proxy: Annotated[FalsyAsNone[HttpURL], Parameter(alias=("http-proxy"))] = None
    "HTTP/HTTPS proxy"

    rate_limit: PositiveFloat = 25
    "Max number of requests per second (only used while scraping)"

    connection_timeout: PositiveFloat = 15
    read_timeout: FalsyAsNone[PositiveFloat] = 300
    ssl_context: FalsyAsNone[
        Annotated[
            Literal["truststore", "certifi", "truststore+certifi"],
            strings.pre_validator(to_lower=True, strip=True),
        ]
    ] = Field(default="truststore+certifi", deprecated=True)
    tls: TLS = Field(default_factory=TLS)

    user_agent: NonEmptyStr = "Mozilla/5.0 (X11; Linux x86_64; rv:153.0) Gecko/20100101 Firefox/153.0"
    impersonate: FalsyAsNone[ImpersonateTarget] = None
    "Use this target as impersonation for all scrape requests"

    @override
    def model_post_init(self, context: Any, /) -> None:
        super().model_post_init(context)
        # https://github.com/FlareSolverr/FlareSolverr/issues/1685
        if self.flaresolverr_concurrency > 1 and self.flaresolverr_use_session:
            raise ValueError("'flaresolverr_concurrency' can't be > 1 if 'flaresolverr_use_session' is True")


class UIMode(CIStrEnum):
    DISABLED = auto()
    ACTIVITY = auto()
    SIMPLE = auto()
    FULLSCREEN = auto()

    @property
    def is_disabled(self) -> bool:
        return self is UIMode.DISABLED

    @property
    def is_fullscreen(self) -> bool:
        return self is UIMode.FULLSCREEN


class UIOptions(ConfigGroup):
    mode: Annotated[UIMode, Parameter(name="--ui")] = UIMode.FULLSCREEN
    portrait: bool = False
    "force CDL to run with a vertical layout"
    refresh_rate: PositiveFloat = 10.0

    show_stats: Annotated[bool, Parameter(name="stats")] = True
    "Show stats report at the end of a run"


class MaxChildren(ConfigGroup, name=None):
    forum: NonNegativeInt = 0
    "Do not scrape more that this number of URLs inside a forum thread"
    forum_post: NonNegativeInt = 0
    "Do not scrape more that this number of URLs inside an individual a forum post"
    profile: NonNegativeInt = 0
    "Do not scrape more that this number of URLs in a profile"
    album: NonNegativeInt = 0
    "Do not scrape more that this number of URLs in a album"
